### DISCUSSION TITLE
Is Heat going away?

### DISCUSSION THREAD

**Discussion by claybrooks** [wixtoolset/issues#8799] [Questions]
Is Heat going to be maintained in v5 and beyond?  Or are you guys moving away from Heat towards the new Files feature introduced in v5, with a plan to deprecate Heat at some point in the future?

**Comment by robmen** [comment]
[Heat is going away](https://github.com/wixtoolset/issues/issues/6838). Probably in v7 or v8.

**Reply by claybrooks** [reply]
Since the WIP for Files on GitHub is closed, I suppose this is the right spot to discuss Files features and usage.

I appreciate the simplicity of Files introduced in v5, and looks to be headed in a good direction. But, obviously there is a very big difference in paradigms between Heat and Files.

- Heat is explicit opt-in while Files is mostly explicit opt-out.  

With the Heat paradigm, I can see unexpected changes from release to release, I get errors at compile time for missing files (broken post build events, missing runtime-only dependencies, weird visual studio behavior, etc.), and I have a strong guarantee that, if my wix project compiles, my MSI is everything that I need it to be, nothing more and nothing less. It's resilient against dirty folders and abnormal filesystem behavior. 

With the current implementation of Files, I lose all of this. I can't see diffs anymore, missing file errors are pushed to install time and runtime instead of compile time, and I actually don't know what's in the packaged MSI unless I crack it open and look at it. It's not resilient at all to dirty folders, as Files doesn't seem to be intimate with any build system.

Am I misunderstanding any of this? If not, are there plans to move Files toward a direction where it provides the strong guarantees we had with   the Heat paradigm?

**Reply by barnson** [reply]
How are you running Heat that it's "explicit opt-in"?

**Reply by claybrooks** [reply]
We run Heat outside of the build process and maintain by "hand". Essentially we have scripts that we run periodically and we perform code reviews before we accept merging in.

It's explicit opt-in because each file must be listed explicitly in wix source files for it to be included in the package.

**Reply by barnson** [reply]
That was the original design of Heat but isn't how most people use it. `Files` replaces the Heat-in-the-build case. There is no current replacement for the generate-then-maintain case but [naked `File`](https://github.com/wixtoolset/issues/issues/7696) and the `Directory` and `Subdirectory` attributes would make it straightforward to do so and make for much better-looking authoring than Heat produced.

**Reply by claybrooks** [reply]
Thank you for the better terminology.  

I definitely think there is a use case for heat-in-the-build, obviously, otherwise you guys wouldn't have done it. I can see it for hobby/small projects and projects that can tolerate rebuilding packages without too much headache.

Generate-and-maintain is more cost effective for highly regulated environments. The cost of building a generate-and-maintain system once is cheap compared to the cost of delivering broken installer packages every now and again. Weeks/months of paperwork when our delivery is broken just isn't worth the sleekness and ease of Files.

Of all the wonderful things that Wix makes easy for developers, acting as a gating mechanism to validate your build outputs is high on the list, for me at least. 

If Heat goes away in v7/v8 with no intent to replace, my guess is something 3rd party will take its place shortly.

**Reply by barnson** [reply]
You're preaching to the choir but Heat is a mess and for years, nobody's volunteered to maintain it.
